File: frontend/src/api/analysis.ts
import request from './request';

export const POINT_INDEX_CODES = [
  '000001',
  '000300',
  '399006',
  '000688',
] as const;

export const DEFAULT_POINT_THRESHOLDS: API.PointThresholds = {
  '000001': 4000,
  '000300': 4500,
  '399006': 2500,
  '000688': 1200,
};

export function isPriceLevelPending(
  item: API.PriceLevelRow,
): item is API.PriceLevelPendingItem {
  return 'data_pending' in item && item.data_pending === true;
}

export function isMarketOverviewError(
  item: API.MarketOverviewRow,
): item is API.MarketOverviewErrorItem {
  return 'error' in item;
}

/** 获取多指数牛市点位达标统计 GET /analysis/bull-markets/point */
export function fetchBullMarketPointStats(
  thresholds: API.PointThresholds,
): Promise<API.MultiIndexPointStats> {
  return request.get('/analysis/bull-markets/point', {
    params: {
      threshold_000001: thresholds['000001'],
      threshold_000300: thresholds['000300'],
      threshold_399006: thresholds['399006'],
      threshold_000688: thresholds['000688'],
    },
  });
}

/** 获取牛市成交额达标统计 GET /analysis/bull-markets/turnover */
export function fetchBullMarketTurnoverStats(
  threshold: number,
): Promise<API.BullMarketStats> {
  return request.get('/analysis/bull-markets/turnover', {
    params: { threshold },
  });
}

/** 获取大盘成交额 TopN GET /analysis/turnover/ranking */
export function fetchTurnoverRanking(
  top: number,
): Promise<API.TurnoverRanking> {
  return request.get('/analysis/turnover/ranking', {
    params: { top },
  });
}

/** 获取个股成交额 TopN GET /analysis/stock/ranking */
export function fetchStockRanking(top: number): Promise<API.StockRanking> {
  return request.get('/analysis/stock/ranking', {
    params: { top },
  });
}

/** 获取全球资产价格水位 GET /analysis/asset-price-levels */
export function fetchAssetPriceLevels(
  forceRefresh = false,
): Promise<API.AssetPriceLevels> {
  return request.get('/analysis/asset-price-levels', {
    params: { force_refresh: forceRefresh || undefined },
  });
}

/** 获取全球市场概览 GET /analysis/market-overview */
export function fetchMarketOverview(
  forceRefresh = false,
): Promise<API.MarketOverview> {
  return request.get('/analysis/market-overview', {
    params: { force_refresh: forceRefresh || undefined },
  });
}

/** 获取美国宏观月度序列 GET /analysis/us-macro */
export function fetchUsMacroData(start = '2020-06'): Promise<API.UsMacroData> {
  return request.get('/analysis/us-macro', {
    params: { start },
  });
}

/** 获取中国宏观月度序列 GET /analysis/cn-macro */
export function fetchCnMacroData(start?: string): Promise<API.CnMacroData> {
  return request.get('/analysis/cn-macro', {
    params: start ? { start } : undefined,
  });
}

File: backend/astock/routers/analysis.py
from fastapi import APIRouter, Query

from astock.config import (
    CN_MACRO_START_PERIOD,
    POINT_INDEX_CONFIG,
    TURNOVER_THRESHOLD,
    US_MACRO_START_PERIOD,
)
from astock.core.deps import DbSession
from astock.schemas.analysis import (
    BullMarketStatsResponse,
    CnMacroResponse,
    MarketOverviewResponse,
    MultiIndexPointStatsResponse,
    PriceLevelsResponse,
    StockRankingResponse,
    TurnoverRankingResponse,
    UsMacroResponse,
)
from astock.schemas.response import ApiResponse, success
from astock.services.global_asset import get_price_levels
from astock.services.market_overview import get_market_overview
from astock.services.queries import (
    bull_market_multi_index_point_stats,
    bull_market_turnover_stats,
    get_cn_macro,
    get_us_macro,
    stock_ranking,
    turnover_ranking,
)

router = APIRouter(prefix="/api/v1/analysis", tags=["analysis"])


@router.get(
    "/bull-markets/point",
    response_model=ApiResponse[MultiIndexPointStatsResponse],
)
def bull_market_point_stats_api(
    db: DbSession,
    threshold_000001: float = Query(
        default=POINT_INDEX_CONFIG["000001"]["default_threshold"],
        gt=0,
    ),
    threshold_000300: float = Query(
        default=POINT_INDEX_CONFIG["000300"]["default_threshold"],
        gt=0,
    ),
    threshold_399006: float = Query(
        default=POINT_INDEX_CONFIG["399006"]["default_threshold"],
        gt=0,
    ),
    threshold_000688: float = Query(
        default=POINT_INDEX_CONFIG["000688"]["default_threshold"],
        gt=0,
    ),
):
    thresholds = {
        "000001": threshold_000001,
        "000300": threshold_000300,
        "399006": threshold_399006,
        "000688": threshold_000688,
    }
    return success(bull_market_multi_index_point_stats(db, thresholds))


@router.get(
    "/bull-markets/turnover",
    response_model=ApiResponse[BullMarketStatsResponse],
)
def bull_market_turnover_stats_api(
    db: DbSession,
    threshold: float = Query(default=TURNOVER_THRESHOLD, gt=0),
):
    return success(bull_market_turnover_stats(db, threshold))


@router.get(
    "/turnover/ranking",
    response_model=ApiResponse[TurnoverRankingResponse],
)
def turnover_ranking_api(
    db: DbSession,
    top: int = Query(default=20, ge=1, le=100),
    bull_market: str | None = Query(default=None),
):
    return success(turnover_ranking(db, top=top, bull_market=bull_market))


@router.get(
    "/stock/ranking",
    response_model=ApiResponse[StockRankingResponse],
)
def stock_ranking_api(
    db: DbSession,
    top: int = Query(default=20, ge=1, le=100),
    bull_market: str | None = Query(default=None),
):
    return success(stock_ranking(db, top=top, bull_market=bull_market))


@router.get(
    "/asset-price-levels",
    response_model=ApiResponse[PriceLevelsResponse],
)
def asset_price_levels_api(
    db: DbSession,
    force_refresh: bool = Query(default=False),
):
    return success(get_price_levels(db, force_refresh=force_refresh))


@router.get(
    "/market-overview",
    response_model=ApiResponse[MarketOverviewResponse],
)
def market_overview_api(
    force_refresh: bool = Query(default=False),
):
    return success(get_market_overview(force_refresh=force_refresh))


@router.get(
    "/us-macro",
    response_model=ApiResponse[UsMacroResponse],
)
def us_macro_api(
    db: DbSession,
    start: str = Query(default=US_MACRO_START_PERIOD, min_length=7, max_length=7),
):
    return success(get_us_macro(db, start=start))


@router.get(
    "/cn-macro",
    response_model=ApiResponse[CnMacroResponse],
)
def cn_macro_api(
    db: DbSession,
    start: str = Query(default=CN_MACRO_START_PERIOD, min_length=7, max_length=7),
):
    return success(get_cn_macro(db, start=start))
